### PR TITLE
Implemented Mimicry

### LINKED ISSUE #1033: ABILITY_MIMICRY

### PR DESCRIPTION
**Pull request by LOuroboros**
## Description
Closes #1033.

## **Discord contact info**
Lunos#4026

### CODE REVIEW

File: data/battle_scripts_1.s [hunk 1]
```diff
@@ -1908,8 +1908,25 @@ BattleScript_EffectPsychicTerrain:
...
+	trytoapplymimicry BS_TARGET, BattleScript_MimicryLoop_NextBattler
+	copybyte gBattlerAbility, sBATTLER
+	call BattleScript_AbilityPopUp
+	printstring STRINGID_BATTLERTYPECHANGEDTO
+	waitmessage B_WAIT_TIME_LONG
+BattleScript_MimicryLoop_NextBattler:
+	addbyte gBattlerTarget, 0x1
+	jumpifbytenotequal gBattlerTarget, gBattlersCount, BattleScript_MimicryLoopIter
+	restoretarget
+	end
```
Comment (ghoulslash) [inline data/battle_scripts_1.s:1928]: this should be `goto BattleScript_MoveEnd` when jumped to from a terrain move.

File: src/battle_util.c [hunk 2]
```diff
@@ -2288,6 +2331,9 @@ u8 DoFieldEndTurnEffects(void)
               && (!(gFieldStatuses & STATUS_FIELD_TERRAIN_PERMANENT) && --gFieldTimers.mistyTerrainTimer == 0))
             {
                 gFieldStatuses &= ~(STATUS_FIELD_MISTY_TERRAIN);
+                for (i = 0; i < MAX_BATTLERS_COUNT; i++)
```
Comment (ghoulslash) [inline src/battle_util.c:2334]: needs braces for the `for` loop

File: src/battle_util.c [hunk 3]
```diff
@@ -2298,8 +2344,12 @@ u8 DoFieldEndTurnEffects(void)
             {
                 if (!(gFieldStatuses & STATUS_FIELD_TERRAIN_PERMANENT)
                   && (gFieldTimers.grassyTerrainTimer == 0 || --gFieldTimers.grassyTerrainTimer == 0))
+                {
                     gFieldStatuses &= ~(STATUS_FIELD_GRASSY_TERRAIN);
-
+                    for (i = 0; i < MAX_BATTLERS_COUNT; i++)
```
Comment (ghoulslash) [inline src/battle_util.c:2349]: braces for the for loop

File: src/battle_util.c [hunk 4]
```diff
@@ -2310,6 +2360,9 @@ u8 DoFieldEndTurnEffects(void)
               && (!(gFieldStatuses & STATUS_FIELD_TERRAIN_PERMANENT) && --gFieldTimers.psychicTerrainTimer == 0))
             {
                 gFieldStatuses &= ~(STATUS_FIELD_PSYCHIC_TERRAIN);
+                for (i = 0; i < MAX_BATTLERS_COUNT; i++)
```
Comment (ghoulslash) [inline src/battle_util.c:2363]: braces. It might also be advantageous to have a single function that does this loop.

**Review by ghoulslash — CHANGES_REQUESTED** [review]

**Review by ghoulslash — APPROVED** [review]

**Review by ghoulslash — CHANGES_REQUESTED** [review]
just realized there is no handling of `BattleScript_EffectRemoveTerrain`

File: data/battle_scripts_1.s [hunk 5]
```diff
@@ -660,6 +660,7 @@ BattleScript_EffectRemoveTerrain:
 	resultmessage
 	waitmessage B_WAIT_TIME_LONG
 	removeterrain
+	trytorevertmimicry
```
Comment (ghoulslash) [inline data/battle_scripts_1.s:663]: why not just include the revert code within `VARIOUS_REMOVE_TERRAIN`?

Comment (LOuroboros) [reply data/battle_scripts_1.s:663]: Didn't think about it too much. I'll do it right away :eyes:

**Comment by LOuroboros** [comment]
I'm using the `for` loop to restore the types of Pokémon with Mimicry 5 times in total.
Should I just pack them in a single function? :eyes:

**Comment by ghoulslash** [comment]
> I'm using the `for` loop to restore the types of Pokémon with Mimicry 5 times in total. Should I just pack them in a single function? 👀

Yeah I think that would be good

**Comment by LOuroboros** [comment]
Ready for re-re-re-re-re-review :eyes:

**Review by ghoulslash — APPROVED** [review]

**Pull request merged**